### PR TITLE
Update CosineAnnealingWarmRestarts to follow PyTorch 1.1+ Step Order.

### LINKED ISSUE #23480: CosineAnnealingWarmRestarts Still Follows PyTorch 1.0 Step Order
## 🐛 Bug

Although the official step order seems to have changed to calling scheduler.step() after each iteration/epoch instead of before, this causes CosineAnnealingWarmRestarts to have an extraneous step at iteration/epoch 0, instead of starting at the maximum specified learning rate. 

## To Reproduce

1. Initialize a scheduler with T_0=1, T_mult=2.
1. Loop through using a for loop, first printing the lr, then stepping.

>>> scheduler = CosineAnnealingWarmRestarts(optimizer, T_0=1, T_mult=2)
>>> for i in range(9):
...          print(i)
...          print(scheduler.get_lr())
...          scheduler.step()
...
0
[0.0]
1
[0.5]
2
[0.5]
3
[0.25]
4
[0.5]
5
[0.42677669529663687]
6
[0.25]
7
[0.07322330470336313]
8
[0.5]

## Expected behavior

For cases such as the one above, would expect the restart to occur at the 8th iteration/epoch (index 7 above), but instead occurs at the 9th iteration/epoch (index 8 above). This is probably due to the resetting of last_epoch in the init, since it resets the last_epoch to the initial user given parameter value, despite the fact that the call to the super() init sets it to user given parameter value + 1.

## Environment

 - PyTorch Version (e.g., 1.0): 1.1
 - OS (e.g., Linux): Windows/WSL
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.7
 - CUDA/cuDNN version: 10
 - GPU models and configuration: GTX 1070
 - Any other relevant information: N/A

## Additional context

N/A

### PR DESCRIPTION
**Pull request by gchanan**
Fixes: https://github.com/pytorch/pytorch/issues/23480.

I only verified that the schedule reaches the restart at the expected step as specified in the issue, it would be good to have someone else verify correctness here.

Script:
```
scheduler = torch.optim.lr_scheduler.CosineAnnealingWarmRestarts(torch.optim.SGD([torch.randn(1, requires_grad=True)], lr=0.5), T_0=1, T_mult=2)
for i in range(9):
    print(i)
    print(scheduler.get_lr())
    scheduler.step()
```
Output:
```
0
[0.5]
1
[0.5]
2
[0.25]
3
[0.5]
4
[0.42677669529663687]
5
[0.25]
6
[0.07322330470336313]
7
[0.5]
8
[0.4809698831278217]
```

### CODE REVIEW

**Comment by ezyang** [comment]
cc @LLYX

**Review by ezyang — APPROVED** [review]
Pre-approving, but I helped Greg construct this patch so I am not a good reviewer for correctness

**Review by vincentqb — APPROVED** [review]
This is a good fix and can be merged as is, thanks @LLYX and @gchanan. To be done here or later:
- [ ] Update the docstring for `CosineAnnealingWarmRestarts` to reflect the correct order.
- [ ] `test_CosineAnnealingWarmRestarts_lr1` will catch this error if the `step` (without `epoch` specified) is moved after the verification (in `_test`).
- [ ] Other tests need to be updated also, since most verify the value after stepping. They pass because the epoch parameter is passed, and overrides the natural stepping. This is fixed in #23306.

By the way, we should not use `get_lr` to get the learning rate, as this will recompute a new value and can lead to unexpected results, see #21800.

```
scheduler = torch.optim.lr_scheduler.CosineAnnealingWarmRestarts(torch.optim.SGD([torch.randn(1, requires_grad=True)], lr=0.5), T_0=1, T_mult=2)
for i in range(9):
    print(i)
    print([param_group['lr'] for param_group in opt.param_groups])
    scheduler.step()
```
